frontends/p4/typeChecking/typeCheckDecl.cpp

```diff
@@ -212,19 +220,38 @@ bool TypeInference::checkAbstractMethods(const IR::Declaration_Instance *inst,
     return rv;
 }
 
-const IR::Node *TypeInference::preorder(IR::Declaration_Instance *decl) {
+TypeInferenceBase::PreorderResult TypeInferenceBase::preorder(
+    const IR::Declaration_Instance *decl) {
     // We need to control the order of the type-checking: we want to do first
     // the declaration, and then typecheck the initializer if present.
-    if (done()) return decl;
-    visit(decl->type, "type");
-    visit(decl->arguments, "arguments");
-    visit(decl->annotations, "annotations");
+    if (done()) return {decl, false};
+    // FIXME: this `cloned` dance is very ugly
+    IR::Declaration_Instance *cloned = nullptr;
+    auto *t = apply_visitor(decl->type, "type")->checkedTo<IR::Type>();
+    if (t != decl->type) {
+        cloned = decl->clone();
+        cloned->type = t;
+        decl = cloned;
+    }
+    auto *args = apply_visitor(decl->arguments, "arguments")->checkedTo<IR::Vector<IR::Argument>>();
```

Comment:
could this be cleaned up by having two preorder methods for Declaration_Instance -- one that takes a const pointer and one that takes a non-const, and then have those call a single template method?  something like:

    template <class T> TypeInferenceBase::PreorderResult TypeInferenceBase::preorder_DeclInst(T decl) {
        // orig code from preorder, unchanged
    }

    TypeInferenceBase::PreorderResult TypeInferenceBase::preorder(const IR::Declaration_Instance *decl) {
        return preorder_DeclInst(decl);
    }
    TypeInferenceBase::PreorderResult TypeInferenceBase::preorder(IR::Declaration_Instance *decl) {
        return preorder_DeclInst(decl);
    }

That way the same code can be used for both the const and non-const versions without all this extra manual cloning ande checking?

Reply:
Thanks. This worked in ~95% of cases, so we ended only of couple of manual clones where we modified internals by ourselves.

Please take a look at https://github.com/p4lang/p4c/pull/4829/commits/720b7b8b8d38deead0a98ae613813f4539ee64ba